mail_embed_image/models/ir_mail_server.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 Therp BV <https://therp.nl>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+import uuid
+import logging
+from contextlib import contextmanager
+from odoo import models, http
+from odoo.addons.base.ir.ir_mail_server import encode_header_param
+from werkzeug.test import EnvironBuilder
+from werkzeug.wrappers import Request as WerkzeugRequest
+from lxml.html.soupparser import fromstring
+from lxml.etree import tostring
+from base64 import encodestring
+import threading
+from odoo.http import root as root_wsgi
+from email.mime.image import MIMEImage
+
+
+logger = logging.getLogger(__name__)
+
+
+class IrMailServer(models.Model):
+    _inherit = 'ir.mail_server'
+
+    @contextmanager
+    def fetch_image(self, path):
+        public_user = self.env.ref('base.public_user')
+        session_store = root_wsgi.session_store
+        session = session_store.new()
+        session.update({
+            'db': threading.current_thread().dbname,
+            'login': public_user.login,
+            'uid': public_user.id,
+            'context': self.env.context,
+            })
+        werkzeug_env = EnvironBuilder(path).get_environ()
+        werkzeug_request = WerkzeugRequest(werkzeug_env)
+        werkzeug_request.session = session
+        # construct an odoo request with this werkzeug request.
+        request = http.HttpRequest(werkzeug_request)
+        with request:
+            request._env = self.env(user=public_user)
+            endpoint, arguments = http.routing_map(
+                self.env.registry._init_modules,
+                False,
+                self.env['ir.http']._get_converters()
+            ).bind_to_environ(
+                werkzeug_env).match(return_rule=False,)
```

Comment:
I'm still not sure if we can keep the reference to the rule, and somehow make it just parse the new arguments, then we'd only have to do the above once. But don't change anything now, let's see if we fine a better way in the future.